File: rust/sleepsort/src/bin/simple.py
import asyncio
from typing import List, Callable

async def sleep_and_do(sleep_seconds: int):
    await asyncio.sleep(sleep_seconds / 1000.)
    return sleep_seconds

async def sort(vals: List) -> List:
    return [await x for x in asyncio.as_completed([sleep_and_do(val) for val in vals])]

async def sort_broken(vals: List) -> List:
    return [x for x in await asyncio.gather(*[sleep_and_do(val) for val in vals])]

async def main():
    original = [10, 0, 2, 6]
    print(f"original: {original}")
    print(f"sorted: {await sort(original)}")
    print(f"sorted: {await sort_broken(original)}")

if __name__ == '__main__':
    asyncio.run(main())

File: rust/sleepsort/src/bin/simple.rs
use std::time::Duration;

use futures::{stream::{FuturesUnordered, FuturesOrdered}, StreamExt};
use tokio::time::sleep;

async fn sleep_and_do<F, T>(sleep_seconds: u64, do_what: F) -> T where F: FnOnce() -> T {
    sleep(Duration::from_millis(sleep_seconds)).await;
    do_what()
}

async fn sort(vals: Vec<u64>) -> Vec<u64>{
    let futs: FuturesUnordered<_> = vals.iter().map(|val| {sleep_and_do(*val, move ||{val})}).collect();
    futs.collect().await
}

async fn sort_broken(vals: Vec<u64>) -> Vec<u64>{
    let futs: FuturesOrdered<_> = vals.iter().map(|val| {sleep_and_do(*val, move ||{val})}).collect();
    futs.collect().await
}


#[tokio::main]
async fn main() {
    let original = vec![10, 0, 2, 6];
    println!("original: {original:?}");
    println!("sort: {:?}", sort(original.clone()).await);
    println!("broken sort: {:?}", sort_broken(original.clone()).await)
}